src/batcher.c

```diff
@@ -0,0 +1,184 @@
+    if (empty_elems == empty_mask) {
+        return false;
+    }
+    iter->elems_to_visit = ~empty_elems & empty_mask;
+    iter->elems_to_reset = 0;
+    return true;
+}
+
+int
+batcher_pop_next(
+  batcher_t *batcher, batcher_elem_t *elems, batcher_pop_iter_t *iter) {
+    while (iter->elems_to_visit != 0) {
+        unsigned idx = ffs_u32(iter->elems_to_visit);
+        uint32_t idx_mask = (uint32_t)1 << idx;
+        iter->elems_to_visit ^= idx_mask;
+        uint8_t elem_state = atomic_load_u8(&elems[idx].state, ATOMIC_ACQUIRE);
+        if (elem_state != batcher_elem_state_full) {
+            // Note: This assert wouldn't be correct: there's a race with push
+            // in which they've claimed a bit but not yet initialized elem_data.
+            // assert(elem_state == batcher_elem_state_pending);
+            continue;
+        }
+        iter->elems_to_reset |= idx_mask;
+        return (int)idx;
+    }
+    return BATCHER_NO_IDX;
+}
+
+void
+batcher_pop_end(
+  batcher_t *batcher, batcher_elem_t *elems, batcher_pop_iter_t *iter) {
+    uint32_t elems_to_reset = iter->elems_to_reset;
+    uint32_t empty_slots_mask = elems_to_reset;
+    while (elems_to_reset != 0) {
+        unsigned idx = ffs_u32(elems_to_reset);
+        uint32_t idx_mask = (uint32_t)1 << idx;
+        elems_to_reset ^= idx_mask;
+        uint8_t elem_state = atomic_load_u8(&elems[idx].state, ATOMIC_ACQUIRE);
+        assert(elem_state == batcher_elem_state_full);
+        atomic_store_u8(
+          &elems[idx].state, batcher_elem_state_empty, ATOMIC_RELEASE);
+    }
+    if (empty_slots_mask != 0) {
+        atomic_fetch_or_u32(
+          &batcher->empty_elems, empty_slots_mask, ATOMIC_RELEASE);
+    }
+}
+
+void
+batcher_prefork(batcher_t *batcher, batcher_elem_t *elems) {
+    (void)batcher;
+    (void)elems;
+}
+
+void
+batcher_postfork_parent(batcher_t *batcher, batcher_elem_t *elems) {
+    (void)batcher;
+    (void)elems;
+}
+
+void
+batcher_postfork_child(batcher_t *batcher, batcher_elem_t *elems) {
+    /*
+     * There are two cases in which we can have push/fork races:
+     * - The pusher is getting its elements from a source protected by a mutex
+     *   with lower rank than the popper.  In this case, we should acquire the
+     *   mutex earlier as part of forking, so no pending pushes should be
+     *   possible.
+     * - The source of elements being pushed is unsynchronized with respect to
+     *   the fork mutexes.  In this case, there's a fundamental risk of losing
+     *   elements (imagine the fork had happens just prior to a
+     *   batcher_push_begin; whatever elements were about to be pushed won't
+     *   make it into shared data structures).  Since this can happen
+     *   regardless, we can just discard the data in any slots that were
+     *   mid-push (because the caller can't tell the difference between a push
+     *   call that didn't quite start and one that started but didn't finish).
```

Comment:
While this seems inevitable, will the data loss here lead to memory leaks?

Reply:
Yes, but that bug is already present. The thing to note is that using the batcher can't introduce any *new* bugs. I beefed up the comment here; PTAL.